dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/tei/query/QueryContext.java

```diff
@@ -0,0 +1,212 @@
+
+import static org.hisp.dhis.analytics.tei.query.QueryContextConstants.ANALYTICS_TEI_TEI;
+import static org.hisp.dhis.analytics.tei.query.QueryContextConstants.TEI_TEI_DEFAULT_ALIAS;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.hisp.dhis.analytics.common.AnalyticsPagingAndSortingParams;
+import org.hisp.dhis.analytics.common.dimension.DimensionIdentifier;
+import org.hisp.dhis.analytics.common.dimension.DimensionParam;
+import org.hisp.dhis.analytics.tei.TeiQueryParams;
+import org.hisp.dhis.analytics.tei.query.items.EventDataValueCondition;
+import org.hisp.dhis.analytics.tei.query.items.OrCondition;
+import org.hisp.dhis.analytics.tei.query.items.Table;
+import org.hisp.dhis.common.BaseIdentifiableObject;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramStage;
+import org.hisp.dhis.program.ProgramTrackedEntityAttribute;
+
+@RequiredArgsConstructor( staticName = "of" )
+public class QueryContext
+{
+    @Getter
+    private final TeiQueryParams teiQueryParams;
+
+    private final ParameterManager parameterManager = new ParameterManager();
+
+    public String getMainTableName()
+    {
+        return ANALYTICS_TEI_TEI + getTeTTableSuffix();
+    }
+
+    public String getTeTTableSuffix()
+    {
+        return teiQueryParams.getTrackedEntityType().getUid();
+    }
+
+    public Query getQuery()
+    {
+        return Query.builder()
+            .select( getSelect() )
+            .from( getFrom() )
+            .where( getWhere() )
+            .order( getOrder() )
+            .limit( getLimit() )
+            .build();
+    }
+
+    private LimitOffset getLimit()
+    {
+        AnalyticsPagingAndSortingParams pagingAndSortingParams = teiQueryParams.getCommonParams()
+            .getPagingAndSortingParams();
+        return LimitOffset.of(
+            pagingAndSortingParams.getPageSize(),
+            pagingAndSortingParams.getPageSize() * pagingAndSortingParams.getPage() );
+    }
+
+    private Order getOrder()
+    {
+        return Order.ofOrder( "" );
+    }
+
+    private Select getSelect()
+    {
+        Select.SelectBuilder builder = Select.builder()
+            // static fields
+            .field( getFieldWithAlias( "trackedentityinstanceid" ) )
+            .field( getFieldWithAlias( "trackedentityinstanceuid" ) );
+
+        // TET and program attribute fields
+        Stream.concat(
+            // Tracked entity Type attributes
+            teiQueryParams.getTrackedEntityType().getTrackedEntityTypeAttributes().stream(),
+            // Program attributes
+            teiQueryParams.getCommonParams().getPrograms().stream()
+                .map( Program::getProgramAttributes )
```

Comment:
Can `getProgramAttributes` be null or contain any null element?

Reply:
nope